src/vm/dllimport.cpp

```diff
@@ -3463,6 +3463,12 @@ BOOL NDirect::MarshalingRequired(MethodDesc *pMD, PCCOR_SIGNATURE pSig /*= NULL*
 
             case ELEMENT_TYPE_VALUETYPE:
             {
+                if (IsReadyToRunCompilation())
+                {
+                    // All sturcts will get marshalling in R2R because structs are loaded at runtime.
```

Comment:
I do not understand this comment. Could you please elaborate?

Reply:
The thing here is that we should be able to handle some structs (ex: blittable, enum, ...), however, what if the shape of the struct changes, and when we load their MethodTables later at runtime, we discover that they are no longer blittable? In that case, the codegen would be wrong because we would require marshalling...

I think a more correct check would have been: if(IsReadyToRunCompilation() && "types used in struct are primitive or in the current module being compiled"). This would ensure that the codegen we generate for a particular R2R image is compatible with the types in that module. But this is a complex check that would require a recursive field checking logic (which I can totally add if it makes sense to have). The scenario I'm trying to avoid is having a struct, where one of the fields is of a type declared in a different module, and that different module gets patched, causing the field in the struct to no longer be blittable for instance.

Does this make sense?

Reply:
Right, it should be sufficient to check that the struct is in the current version bubble.

Reply:
I think it's more than checking if the struct is in the current bubble. The struct could have fields of other struct types that could be outside of the bubble, so we should recursively check the struct and all of its fields